### ISSUE TITLE
Allow gridmap and tilemap items to span multiple cells

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#1713]
**Describe the project you are working on:**
3d Rougelike

**Describe the problem or limitation you are having in your project:**
Currently I need to have 2 tall blocks for walls for a flat 3D dungeon so implementing AStar pathing is easier.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Can tilemaps and gridmaps have the option for items to span multiple cells?
This would allow us to have also create 3d builders without needing to check cell sizes manually

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Here is the issue its based on https://github.com/godotengine/godot/issues/17921
I'm not sure what the best way to implement this is but ideally we would need to be able to specify a item size in cells.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
it can be worked around with multipule tilesets or meshlibraries but thats kinda a hack 

**Is there a reason why this should be core and not an add-on in the asset library?:**
right now gridmaps are pretty barebones and I would like to see more features go into them

**Comment by Calinou** [comment]
This should already be feasible by creating a mesh/texture that is larger than its containing cell and placing it in the GridMap/TileMap.

**Comment by Shadowblitz16** [comment]
@Calinou the issue is the current solution allows for overlapping tiles

**Comment by ATHellboy** [comment]
Is there anyway to find which cells are occupied by the large item ?
For instance the grid map cell size is 2x2x2 and there is an item like this one:
![gridmap](https://user-images.githubusercontent.com/10496528/145201226-9dc2a5f7-21b5-4207-8e26-c9bf24b4bde9.png)
Can figure out which cells are empty and which ones are full ? Because GridMap itself just show the pivot point as a full cell.

**Comment by twoco** [comment]
Use case: A mesh lib with different sized objects in grid like 1x1, 2x2, 3x3, etc. For example: Have a road/trail of 1x1 and a house 3x3 like in a city building game e.g. Anno series.

**Comment by Calinou** [comment]
> This is not possible in current GridMap?

As described above, you can already have meshes that are larger than the cell they are placed in. You will need to perform custom checks in your code to prevent users (or your own generation code) from placing overlapping tiles that don't make sense (e.g. a house within another).

**Comment by twoco** [comment]
Right, sorry. Forgot to apply transformation in Blender (CTRL+A) before export for the test. ^^ Ok, it works. Not perfect yet as the pivot of the objects seems to be ignored and are not correctly placed on the map. But not sure what the problem is. I will find out and possibly report back. To this issue: 3D tilemaps works a bit different like 2D tilemaps. On a pixel map each cell has exactly the size (e.g. 16x16px). If an object like a building needs 32x16 you draw it on 2x1 tiles. On 3D tile maps, your objects may not exactly match the grid size (e.g. 1x1). The map does not know the size of the object. Like a 2D tilemap does not know the size of the objects drawn on the bitmap. If we want exactly the same behavior like on a 2D tilemap, we need to split the larger object (2x1) to 1x1 tiles. But it may not always be a pleasure to do that. The only thing I think spontaneously is wrapping each object in a grid scope. Means you place a 3D model in cell context. Like if you select all tiles on a 2D map and save it as object, so the map know the size what number of cells it will take. But idk, don't know how that would look like implemented. Anyway, it's always for developer in Godot. For users or map generation you (currently) need to implement this by your own. E.g. creating objects of your tiles which the required data like size and rotation etc. Even if the Godot map can handle, you have to handle that collisions (replacements). Anyway, first I need to solve my issues with the pivot. Godot seems also have problems with the cache when replacing objects (actual objects not updated/used. Only the already deleted).

You can try it out: Simply create 2x MeshInstances (or in Blender with custom origin):
- Road: Cube X: 1 Y: 0.1 Z: 1
- House: Cube X: 5 Y: 2 Z: 4
Save it as mesh library and create a GridMap with cell 1,1,1 and all center options disabled. Place a road and house and play with that options. I cannot place these objects edge on edge. Even if I use objects from blender with origin 0,0,0 (not centered like the MeshInstance). Maybe I need more investigation to it. And maybe write my own grid system. But hints or examples are appreciated.

[GridMapDemo.zip](https://github.com/godotengine/godot-proposals/files/10335289/GridMapDemo.zip)

<img src="https://user-images.githubusercontent.com/53160891/210326813-b4a5d034-ed27-4472-b2e3-b3bd5960acef.jpg" width="200"/>

**Comment by twoco** [comment]
Right, sorry. Forgot to apply transformation in Blender (CTRL+A) before export for the test. ^^ Ok, it works. For some sizes I need to move the origin axis to 0.5. Otherwise it's impossible to place the objects edge on edge. In the image below you see a road with x1z1 tiles and a building x4z5. I have to move the origin on x axis to 0.5 and translate back to zero (in Blender). Otherwise you'll have a gap. But anyway, I found out how to work with. It's a great tool in Godot. 

<img src="https://user-images.githubusercontent.com/53160891/210367689-3caf62ef-63f8-40c6-9388-7db4c6a7fdd6.jpg" width="200" />

To this issue: The GridMap (3D tilemap) works similar as a 2D tilemap. Both systems do not know the size of the objects. Yes, a 2D tilemap has a fix size of e.g. 16x16px. But this is similar to the cell size of GridMap. A 2D tilemap also does not know what's on the bitmap. It does not know what you have drawn. E.g. a house over 2x3 tiles. You just don't thinking on it because the house is splitted in 6 tiles, but you see it at one object. If you want the same behavior for GridMap you must split the 3D house model in 6 tiles. Fortunately you don't have to. But it would be great if we find a way to tell the GridMap the size of a tile. It's like selecting all 6 tiles on 2D tilemap and save it as object (group). For the GridMap we must tell that object that it takes 6 tiles (2x3) in this example. I have no idea how to implement it in Godot. Maybe via the MeshLibary. There you can set the ID, Name and other attributes. We could add an input to enter the cell size (no need to know to the GridMap cell size. It's just a factor and must be an integer. ActualSize = GridMapCellSize * MeshCellSize). On GridMap it will know that the object takes that size. One use case is deleting larger (greater than cell size) object. Because currently it's not that easy to delete such an tile from the map with the right mouse click. It's hard to finde the magic point, you know. ^^ Because it's actually the single cell somewhere inside the bigger tile. With this feature you can just right click everywhere withing the object and it just deletes without hasle. This must also work with rotation. Not sure if this can be easily implemented as addition to the existing GridMap without breaking changes. Just some random thoughts.

**Comment by insomniacUNDERSCORElemon** [comment]
Even a very-limited bitwise system would be nice. For example sizes 1x1x1 up-to 8x8x8\* and only convex tiles (bounding box\*\*, detect first and last cell).

Furthering this would be cell types for placement restriction (wall, floor, etc) and autorotation (for wall cells) and autotiling (9patchrect or simpler with rotation). Multiple gridmaps would still be useful for massive scale difference as well as non-full cells (like type-a and type-b so non-overlapping items can exist in the same cell) so long as there is some control over cell availability\*\*\*

Even better if tile types and their rotation behavior (wall or on_wall has 1 rotation, cube has no rotation) could be added to the Blender node name importing (like -col).

\* including non-cube like 1x2x1

\*\* so an L-shape would either be a full 3x3 (which can work for an interior hallway) or it'd need to be done differently (corner tile and hall tile).

\*\*\* small cells should never be placed inside a cell occupied by a larger tile, it could go the same way in reverse (or the large tile could remove the small cell, or overlapping cells could be detected and hidden for later cleanup/reversal)
